python/packages/autogen-ext/src/autogen_ext/code_executors/jupyter/_jupyter_code_executor.py

```diff
@@ -0,0 +1,185 @@
+import asyncio
+import base64
+import json
+import re
+import sys
+import uuid
+from dataclasses import dataclass
+from pathlib import Path
+from types import TracebackType
+
+if sys.version_info >= (3, 11):
+    from typing import Self
+else:
+    from typing_extensions import Self
+
+from autogen_core import CancellationToken
+from autogen_core.code_executor import CodeBlock, CodeExecutor, CodeResult
+from nbclient import NotebookClient
+from nbformat import NotebookNode
+from nbformat import v4 as nbformat
+
+from .._common import silence_pip
+
+
+@dataclass
+class JupyterCodeResult(CodeResult):
+    """A code result class for Jupyter code executor."""
+
+    output_files: list[Path]
+
+
+class JupyterCodeExecutor(CodeExecutor):
+    def __init__(
+        self,
+        kernel_name: str = "python3",
+        timeout: int = 60,
+        output_dir: Path = Path("."),
+    ):
+        """A code executor class that executes code statefully using nbclient.
+
+        Args:
+            kernel_name (str): The kernel name to use. By default, "python3".
+            timeout (int): The timeout for code execution, by default 60.
+            output_dir (Path): The directory to save output files, by default ".".
+        """
+        if timeout < 1:
+            raise ValueError("Timeout must be greater than or equal to 1.")
+
+        self._kernel_name = kernel_name
+        self._timeout = timeout
+        self._output_dir = output_dir
+        # TODO: Forward arguments perhaps?
+        self._client = NotebookClient(
```

Comment:
Curious is it possible to preload modules and functions to this? As a follow up.

Reply:
Can you elaborate? What kind of functions do you want to preload? I am also not sure I understand preload in this context. The constructor is only executed once, so isn't this already preloading?
Or do you mean moving functionality from here inside the Notebook Client e.g. extend it, monkey patch like?

Reply:
I meant something like allowing pre-defined functions to be added to the kernel so an agent can make use of existing code in the code executor.

Reply:
This can be easily be achieved by just executing code that defines these functions or imports these functions from  a module though. I use this pattern in my code and I think it already is quite simple.